### ISSUE TITLE
istiod high memory usage when create thousands of ServiceEntries

### ISSUE THREAD

**Issue by dozer47528** [istio/istio#25531]
**Bug description**

I created thousands of ServiceEntries in istio and memory usage is very high.

Number of ServiceEntries: 5000
Endpoints in each ServiceEntry: 10

![image](https://user-images.githubusercontent.com/1676951/87530259-ac9ac280-c6c2-11ea-9e0d-2d660f1022f8.png)

**Affected product area (please put an X in all that apply)**

[x] Performance and Scalability

**Expected behavior**

less memory usage

**Steps to reproduce the bug**

[se.yaml.zip](https://github.com/istio/istio/files/4924323/se.yaml.zip)

```bash
kubectl apply -f se.yaml
```

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

client version: 1.6.5
control plane version: 1.6.5
data plane version: 1.6.5

**How was Istio installed?**

```sh
istioctl operator init

kubectl create namespace istio-system

kubectl label namespace default istio-injection=enabled

kubectl apply -f - <<EOF
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: example-istiocontrolplane
spec:
  profile: demo
EOF
```

**Environment where bug was observed (cloud vendor, OS, etc)**

minikube in Ubuntu 20.04

**Comment by dozer47528** [comment]
After debugging the code, I find the root cause of high memory usage.

<img width="1041" alt="Screen Shot 2020-07-15 at 13 58 13" src="https://user-images.githubusercontent.com/1676951/87531449-4e6edf00-c6c4-11ea-8ee7-fd5db7582908.png">

First of all, the total items in all configDependencies will be O(n^2) of the number ServiceEntries, it's too large. We can have a better design to make the items O(n).

On the other hand, the `ConfigKey` is too large, we can replace it with hashcode.

**Comment by dozer47528** [comment]
the root cause was fixed in #25118

**Comment by YonkaFang** [comment]
Considering that the `O(n^2)` complexity problem has been resolve by #25118 , it seems not that worthy to losing the exact dependency info which may be used in some scenarios.
